apollo-federation/src/query_graph/build_query_graph.rs

```diff
@@ -1474,6 +1482,131 @@ impl FederatedQueryGraphBuilder {
         Ok(())
     }
 
+    fn handle_context(&mut self) -> Result<(), FederationError> {
+        let mut subgraph_to_args: IndexMap<Arc<str>, Vec<ObjectFieldArgumentDefinitionPosition>> =
+            IndexMap::default();
+        let mut coordinate_map: IndexMap<
+            Arc<str>,
+            IndexMap<ObjectFieldDefinitionPosition, Vec<ContextCondition>>,
+        > = IndexMap::default();
+        for (subgraph_name, subgraph) in self.base.query_graph.subgraphs() {
+            let subgraph_data = self.subgraphs.get(subgraph_name)?;
+            let Some((_, context_refs)) = &subgraph
+                .referencers()
+                .directives
+                .iter()
+                .find(|(dir, _)| **dir == subgraph_data.context_directive_definition_name)
+            else {
+                continue;
+            };
+
+            let Some((_, from_context_refs)) = &subgraph
+                .referencers()
+                .directives
+                .iter()
+                .find(|(dir, _)| **dir == subgraph_data.from_context_directive_definition_name)
+            else {
+                continue;
+            };
+
+            // Collect data for @context
+            let mut context_name_to_types: HashMap<&str, HashSet<&ObjectTypeDefinitionPosition>> =
+                HashMap::default();
+            for object_def_pos in &context_refs.object_types {
+                let object = object_def_pos.get(subgraph.schema())?;
+                for dir in object.directives.get_all(&CONTEXT_DIRECTIVE_NAME) {
+                    let application = FederationSpecDefinition::context_directive_arguments(dir)?;
+                    context_name_to_types
+                        .entry(application.name)
+                        .or_default()
+                        .insert(object_def_pos);
+                }
+            }
+
+            // Collect data for @fromContext
+            let coordinate_map = coordinate_map.entry(subgraph_name.clone()).or_default();
+            for object_field_arg in &from_context_refs.object_field_arguments {
```

Comment:
It's nice that the way we've architected this, we don't have to do some of the checking that's done in the JS code.